pyquil/api/_quantum_computer.py

```diff
@@ -671,3 +695,340 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    The asymmetry comes from the fact that p00 does not have to be equal to p10 etc.
+
+    :param p00: the probablity of |0,0> to remain in |0,0>
+    :param p01: the probablity of |0,1> to remain in |0,1>
+    :param p10: the probablity of |1,0> to remain in |1,0>
+    :param p11: the probablity of |1,1> to remain in |1,1>
+    :returns: a list of four Kraus operators.
+    """
+    p00e = 1.0 - p00
+    p01e = 1.0 - p01
+    p10e = 1.0 - p10
+    p11e = 1.0 - p11
+    kI = np.array([[np.sqrt(1 - p00e), 0.0, 0.0, 0.0], [0.0, np.sqrt(1 - p01e), 0.0, 0.0],
+                   [0.0, 0.0, np.sqrt(1 - p10e), 0.0], [0.0, 0.0, 0.0, np.sqrt(1 - p11e)]])
+    k00 = np.sqrt(p00e / 3) * (_flip_matrix(0, 1) + _flip_matrix(0, 2) + _flip_matrix(0, 3))
+    k01 = np.sqrt(p01e / 3) * (_flip_matrix(1, 0) + _flip_matrix(1, 2) + _flip_matrix(1, 3))
+    k10 = np.sqrt(p10e / 3) * (_flip_matrix(2, 0) + _flip_matrix(2, 1) + _flip_matrix(2, 3))
+    k11 = np.sqrt(p11e / 3) * (_flip_matrix(3, 0) + _flip_matrix(3, 1) + _flip_matrix(3, 2))
+    return kI, k00, k01, k10, k11
+
+
+def _flip_matrix(i, j, dim=4):
+    mat = np.zeros((dim, dim))
+    # mat.itemset((i,j),1)
+    mat.itemset((j, i), 1)
+    return mat
+
+
+def append_kraus_to_gate(kraus_ops, g):
+    """
+    Follow a gate `g` by a Kraus map described by `kraus_ops`.
+
+    :param list kraus_ops: The Kraus operators.
+    :param numpy.ndarray g: The unitary gate.
+    :return: A list of transformed Kraus operators.
+    """
+    return [kj.dot(g) for kj in kraus_ops]
+
+
+II_mat = np.eye(4)
+II_definition = DefGate("II", II_mat)
+II = II_definition.get_constructor()
+kraus_ops = two_qubit_bit_flip_operators(0.7, 1, 1, 1)
+# will delete code above here
+# ================================================================================================
+
+
+def _flip_array_to_prog(flip_array: Tuple[bool], qubits: List[int]) -> Program:
+    """
+    Generate a pre-measurement program that flips the qubit state according to the flip_array of
+    bools.
+
+    This is used, for example, in symmetrization to produce programs which flip a select subset
+    of qubits immediately before measurement.
+
+    :param flip_array: tuple of booleans specifying whether the qubit in the corresponding index
+        should be flipped or not.
+    :param qubits: list specifying the qubits in order corresponding to the flip_array
+    :return: Program which flips each qubit (i.e. instructs RX(pi, q)) according to the flip_array.
+    """
+    assert len(flip_array) == len(qubits), "Mismatch of qubits and operations"
+    prog = Program()
+    for qubit, flip_output in zip(qubits, flip_array):
+        if flip_output == 0:
+            continue
+        elif flip_output == 1:
+            prog += Program(RX(pi, qubit))
+        else:
+            raise ValueError("flip_bools should only consist of 0s and/or 1s")
+    return prog
+
+
+def _symmetrization(program: Program, meas_qubits: List[int], symm_type: int = 3) \
```

Comment:
my gut says this code doesn't make sense in the QC module, but I need to think about where it belongs

Reply:
Agreed. I wasn't sure where to put it.